rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+a significant performance impact given the relatively broad spans we're
+recommending (i.e. no high-throughput, per-event spans).
+
+One final concern when instrumenting with spans is the length of the span's
+lifetime. In early experiments, we ran into issues where extremely long-lived
+spans (e.g. covering the full runtime of a component) are opened and collect
+data continuously, never flushing because the span does not end. This lead to
+extreme amount of memory growth. The problem was greatly exaggerated by enabling
+`TRACE`-level logging, which resulted in an enormous number of log events
+collecting within the span. To address this, **we proposing pushing the
+outermost spans down from the lifetime of a component to an "iteration" of
+a component**. This will depend greatly on the specifics of each component, but
+the file source above can serve as an example. We also propose keeping these
+spans at the `INFO` level, where our logging is actually pretty quiet and there
+is far less potential for accumulating memory even if spans last longer than
+intended.
+
+In summary:
+
+- Spans should be component-focused, breaking down their operations by stages with at least two levels of hierarchy but up to three levels for components with more complex internals:
+    1. Single "iteration" of operation, on the order of seconds, as a baseline
+    2. High level stages of operation within an iteration, such as reading,
+       processing, sending
+    3. Interesting sub-stages of each operation, with a focus on units of work
+       with significant off-CPU time
+- Spans should be instantiated at the `INFO` log level
+
+#### Forwarding
+
+For the mechanics of actually sending span data to Datadog, we propose relying
+on a couple of existing libraries to handle virtually all of the heavy lifting.
+It's possible that we'll want to extend or replace them in the future, but
+currently the [`tracing-opentelemetry`](https://docs.rs/tracing-opentelemetry/)
+and [`opentelemetry-datadog`](https://docs.rs/opentelemetry-datadog) crates
+combine to fulfill this function to a satisfactory degree.
+
+## Rationale
+
+### Why is this change worth it?
+
+With a relatively small investment of effort, the system described here would be
+able to deliver significant, Vector-specific insight into performance directly
+from deployed environments. The type of instrumentation needed is already
+present within Vector, just underutilized, and it very neatly fills a gap
+between high-level metrics and low-level CPU profiling.
+
+### What is the impact of not doing this?
+
+Performance investigations would continue to rely on a combination of intuition
+and expert-level tooling to narrow down problems reported by users. Instead of
+having data directly indicating that, for example, a component is spending all
+of its time fingerprinting files, we would need to deduce that from other
+sources of information.
```

Comment:
How would this experience differ -- where we request users to send us traces -- from requesting users send us sampled CPU profiles? Both point to where vector was a time hog and require some knowledge of vector internals to interpret / act on.